src/react-disposable-decorator.js

```diff
@@ -18,30 +18,52 @@ export default function(DecoratedComponent) {
     constructor(props) {
       super(props);
       this.disposables = [];
+      this.mounted = true;
     }
 
     render() {
       return <DecoratedComponent {...this.props} cancelWhenUnmounted={this.cancelWhenUnmounted} cancelAllSubscriptions={this.cancelAllSubscriptions} />
     }
     componentWillUnmount() {
       this.cancelAllSubscriptions();
+      this.mounted = false;
     }
     cancelWhenUnmounted = (...thingsToCancel) => {
+      if (thingsToCancel.length === 0) {
+        throw new Error(`cancelWhenUnmounted should be called with one or more cancelables (an object with a dispose, unsubscribe, or cancel function)`);
+      }
+
       thingsToCancel.forEach(thingToCancel => {
-        if (!thingToCancel || typeof thingToCancel.dispose !== 'function') {
-          throw new Error(`cancelWhenUnmounted should be called with one or more disposables (an object with a dispose function)`);
+        if (!thingToCancel || (typeof thingToCancel.dispose !== 'function' && typeof thingToCancel.cancel !== 'function' && typeof thingToCancel.unsubscribe !== 'function')) {
+          throw new Error(`cancelWhenUnmounted should be called with one or more cancelables (an object with a dispose, unsubscribe, or cancel function)`);
+        }
+
+        if (this.mounted) {
+          this.disposables.push(thingToCancel);
+        } else {
+          // They called cancelWhenUnmounted after the component unmounted...
+          cancel(thingToCancel)
```

Comment:
immediately cancel a subscription if cancelWhenUnmounted is called after the component is already unmounted.